lib/Runtime/Base/ObjectTypeCache.cpp

```diff
@@ -0,0 +1,51 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeBasePch.h"
+#include "Types/PathTypeHandler.h"
+
+namespace Js
+{
+
+DynamicType* ObjectTypeCache::FindOrCreateType(_In_ RecyclableObject* prototype)
+{
+    // For fast lookup, cache index is a function of prototype address
+    const uint index = (uint)(((uintptr_t)prototype) >> PolymorphicInlineCacheShift) & (ObjectTypeCache::MaxCachedTypes - 1);
```

Comment:
Have you collected any stats on how frequently this causes collisions? If prototype objects are similarly sized/allocated from the same heap block or so, we would see excessive collisions here?